File: dart/Generate_Parentheses.dart
/* 22. Generate Parentheses

Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.*/

//solution

class Solution {
  List<String> generateParenthesis(int n) {
    List<String> result = [];
    _backtrack(result, "", 0, 0, n);
    return result;
  }

  void _backtrack(List<String> result, String currentString, int open, int close, int max) {
     if (currentString.length == max * 2) {
      result.add(currentString);
      return;
    }

     if (open < max) {
      _backtrack(result, currentString + "(", open + 1, close, max);
    }

     if (close < open) {
      _backtrack(result, currentString + ")", open, close + 1, max);
    }
  }
}

File: javascript/Generate_Parentheses.js
/* 22. Generate Parentheses

Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

 

Example 1:

Input: n = 3
Output: ["((()))","(()())","(())()","()(())","()()()"]
Example 2:

Input: n = 1
Output: ["()"]*/

//solution:

var generateParenthesis = function(n) {
    const result = [];
    
     function backtrack(currentString, openCount, closeCount) {
         if (currentString.length === n * 2) {
            result.push(currentString);
            return;
        }
        
         if (openCount < n) {
            backtrack(currentString + "(", openCount + 1, closeCount);
        }
        
         if (closeCount < openCount) {
            backtrack(currentString + ")", openCount, closeCount + 1);
        }
    }
    
     backtrack("", 0, 0);
    return result;
};
